File: api/titles.py
import re

_HEADING_RE = re.compile(r"^\s*#{1,6}\s+(.*?)\s*#*\s*$")

_IMAGE_RE = re.compile(r"!\[([^\]]*)\]\([^)]*\)")
_LINK_RE = re.compile(r"\[([^\]]+)\]\([^)]*\)")
_CODE_RE = re.compile(r"`([^`]+)`")
_EMPHASIS_RE = re.compile(r"(\*\*\*|\*\*|\*|___|__|_)(.+?)\1")
_STRIKE_RE = re.compile(r"~~(.+?)~~")


def _strip_inline_markdown(text: str) -> str:
    """Reduce inline Markdown to plain text (bold, italic, links, code, ...)."""
    text = _IMAGE_RE.sub(r"\1", text)
    text = _LINK_RE.sub(r"\1", text)
    text = _CODE_RE.sub(r"\1", text)
    text = _EMPHASIS_RE.sub(r"\2", text)
    text = _STRIKE_RE.sub(r"\1", text)
    return text.strip()


def extract_title(content: str) -> str:
    """Derive a plain-text note title from Markdown text.

    Returns the text of the first ATX heading (``#`` .. ``######``). If there is
    no heading, falls back to the first non-empty line. Inline formatting (bold,
    italic, links, code) is stripped to plain text. If the text is empty or only
    whitespace, returns a placeholder.
    """
    lines = content.splitlines()

    for line in lines:
        match = _HEADING_RE.match(line)
        if match:
            return _strip_inline_markdown(match.group(1))

    for line in lines:
        if line.strip():
            return _strip_inline_markdown(line)

    return "(bez názvu)"

File: frontend/src/titles.ts
// Derives a note's title from its markdown — client-side version (the
// server never sees the content). Must match the earlier backend behavior.
import { translate } from './i18n'

const IMAGE_RE = /!\[([^\]]*)\]\([^)]*\)/g
const LINK_RE = /\[([^\]]+)\]\([^)]*\)/g
const CODE_RE = /`([^`]+)`/g
const EMPHASIS_RE = /(\*\*\*|\*\*|\*|___|__|_)(.+?)\1/g
const STRIKE_RE = /~~(.+?)~~/g

// Strips inline markdown (bold, italic, links, images, code) down to plain text.
export function stripInlineMarkdown(text: string): string {
  return text
    .replace(IMAGE_RE, '$1')
    .replace(LINK_RE, '$1')
    .replace(CODE_RE, '$1')
    .replace(EMPHASIS_RE, '$2')
    .replace(STRIKE_RE, '$1')
    .trim()
}

const HEADING_RE = /^\s*#{1,6}\s+(.*?)\s*#*\s*$/

// The first ATX heading, else the first non-blank line, else a placeholder.
export function extractTitle(content: string): string {
  const lines = content.split('\n')
  for (const line of lines) {
    const m = line.match(HEADING_RE)
    if (m) return stripInlineMarkdown(m[1])
  }
  for (const line of lines) {
    if (line.trim()) return stripInlineMarkdown(line)
  }
  return translate('common.untitled')
}

// Use the given title if any, else derive it from the content (never empty).
export function resolveTitle(title: string | undefined, content: string): string {
  if (title && title.trim()) return title.trim()
  return extractTitle(content)
}
